trin-core/src/portalnet/find/iterators/findnodes.rs

```diff
@@ -0,0 +1,701 @@
+                    Some(peer.key.into_preimage())
+                } else {
+                    None
+                }
+            })
+            .take(self.config.num_results)
+            .collect()
+    }
+
+    /// Checks if the query is at capacity w.r.t. the permitted parallelism.
+    ///
+    /// While the query is stalled, up to `num_results` parallel requests
+    /// are allowed. This is a slightly more permissive variant of the
+    /// requirement that the initiator "resends the FIND_NODE to all of the
+    /// k closest nodes it has not already queried".
+    fn at_capacity(&self) -> bool {
+        match self.progress {
+            QueryProgress::Stalled => self.num_waiting >= self.config.num_results,
+            QueryProgress::Iterating { .. } => self.num_waiting >= self.config.parallelism,
+            QueryProgress::Finished => true,
+        }
+    }
+}
+
+/// Stage of the query.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+enum QueryProgress {
+    /// The query is making progress by iterating towards `num_results` closest
+    /// peers to the target with a maximum of `parallelism` peers for which the
+    /// query is waiting for results at a time.
+    ///
+    /// > **Note**: When the query switches back to `Iterating` after being
+    /// > `Stalled`, it may temporarily be waiting for more than `parallelism`
+    /// > results from peers, with new peers only being considered once
+    /// > the number pending results drops below `parallelism`.
+    Iterating {
+        /// The number of consecutive results that did not yield a peer closer
+        /// to the target. When this number reaches `parallelism` and no new
+        /// peer was discovered or at least `num_results` peers are known to
+        /// the query, it is considered `Stalled`.
+        no_progress: usize,
+    },
+
+    /// A query is stalled when it did not make progress after `parallelism`
+    /// consecutive successful results (see `on_success`).
+    ///
+    /// While the query is stalled, the maximum allowed parallelism for pending
+    /// results is increased to `num_results` in an attempt to finish the query.
+    /// If the query can make progress again upon receiving the remaining
+    /// results, it switches back to `Iterating`. Otherwise it will be finished.
+    Stalled,
+
+    /// The query is finished.
+    ///
+    /// A query finishes either when it has collected `num_results` results
+    /// from the closest peers (not counting those that failed or are unresponsive)
+    /// or because the query ran out of peers that have not yet delivered
+    /// results (or failed).
+    Finished,
+}
+
+/// Representation of a peer in the context of a query.
+#[derive(Debug, Clone)]
+struct QueryPeer<TNodeId> {
+    /// The `KBucket` key used to identify the peer.
+    key: Key<TNodeId>,
+
+    /// The current rpc request iteration that has been made on this peer.
+    iteration: usize,
```

Comment:
Why do we have `iteration` field? It looks like this is not a field for `QueryPeer` on the master branch of `sigp/discv5`. Also, we seem to never use this field.